### PR TITLE
feat: testimonial - delete testimonial

### PR DESCRIPTION
**Pull request by Daniel-Ojo-Williams**
## What does the PR do?

This PR introduces an endpoint allowing Admins to delete user-created testimonials. The implementation includes changes in the Testimonial controller, Testimonial Service, and accompanying tests to ensure proper functionality.

### Changes Implemented
1. **Testimonial Controller**:
   - Added a new endpoint for deleting testimonials.
   -  *Ensured only Admins can access this endpoint through proper authentication and authorization checks. [Waiting on the implementation for this]*

2. **Testimonial Service**:
   - Implemented a service method to handle the deletion logic.
   - Checked for the existence of the testimonial and handled potential errors.

3. **Tests**:
   - Created unit tests for the new service method to ensure it behaves as expected.
   
### Endpoints
- **DELETE /api/testimonials/:id**: Deletes the testimonial with the specified ID. Only accessible to Admins.

Some requirements include:

- [x] User is logged in.
- [ ] User is an admin (Waiting on the feature for this to implement this)
- [x] The testimonial to be deleted exists.

How should this be manually tested?

1. Login to a user account, and get the access token

2. Make a DELETE request to the `\api\testimonials\:id` endpoint with the following payload:

Successful Response

On a successful testimonial creation, the API should return a 201 CREATED status code.

json
{
      "success": "true",
      "message": "Testimonial deleted successfully"
}

Authentication Error:

When the current user is not properly authenticated to change the organization detail:

json
{
  "success": "false",
  "message": "User is currently unauthorized, kindly authenticate to continue",
  "status_code": 401
}

Testimonial not found

On Invalid ID error, if the testimonial is not found with the provided id:

json
{
  "success": "false",
  "message": "Testimonial not found",
  "status_code": 404
}

Task

- [x] Create a DELETE endpoint DELETE /api/v1/testimonials/:id to create testimonials for logged in users.
- [x] Implement server-side logic to handle successful and failed update request.
- [x] Ensure the user has the proper authentication to delete testimonials.
- [x] Write tests for deleting testimonials.
- [x] Update API documentation to reflect the new endpoint and its usage.

Reference Issue: [Link to Issue](https://github.com/hngprojects/hng_boilerplate_nestjs/issues/238)

### CODE REVIEW

File: db/migrations/1721779124878-migration.ts
Comment (CFCIfe) [inline db/migrations/1721779124878-migration.ts:1]: Please remove this file

File: src/main.ts [hunk 1]
```diff
@@ -42,14 +42,11 @@ async function bootstrap() {
     .build();
 
   const document = SwaggerModule.createDocument(app, options);
-  SwaggerModule.setup('api/docs', app, document);
+  SwaggerModule.setup('api', app, document);
```
Comment (CFCIfe) [inline src/main.ts:45]:
```suggestion
SwaggerModule.setup('api/docs', app, document);
```

File: src/main.ts [hunk 2]
```diff
@@ -42,14 +42,11 @@ async function bootstrap() {
...
 
   const port = app.get<ConfigService>(ConfigService).get<number>('server.port');
   await app.listen(port);
 
-  logger.log({ message: 'server started 🚀', port, url: `http://localhost:${port}/api/v1` });
+  logger.log({ message: 'server started 🚀', port, url: `http://localhost:${port}/api` });
```
Comment (CFCIfe) [inline src/main.ts:50]:
```suggestion
  logger.log({ message: 'server started 🚀', port, url: `http://localhost:${port}/api/v1` });
}
bootstrap().catch(err => {
  console.error('Error during bootstrap', err);
  process.exit(1);
});
```

File: src/modules/testimonials/testimonials.service.ts [hunk 3]
```diff
@@ -50,4 +50,16 @@ export class TestimonialsService {
...
+  async deleteTestimonial(id: string) {
+    const testimonial = await this.testimonialRepository.findOneBy({ id });
+
+    if (!testimonial)
+      throw new NotFoundException({
+        success: false,
+        message: 'Testimonial not found',
+        status_code: HttpStatus.NOT_FOUND,
+      });
+    await this.testimonialRepository.delete({ id });
```
Comment (CFCIfe) [inline src/modules/testimonials/testimonials.service.ts:63]: Please implement [onDelete option](https://orkhan.gitbook.io/typeorm/docs/relations#relation-options) in the `./entities/testimonials.entity` file.

File: db/migrations/1721779124878-migration.ts [hunk 4]
```diff
@@ -0,0 +1,23 @@
+import { MigrationInterface, QueryRunner } from 'typeorm';
+
+export class Migration1721779124878 implements MigrationInterface {
```
[the remark this answers is not in the document]
Comment (Daniel-Ojo-Williams) [reply db/migrations/1721779124878-migration.ts]: Done, removed the migration file

File: db/migrations/1721779124878-migration.ts
Comment (Daniel-Ojo-Williams) [reply db/migrations/1721779124878-migration.ts:1]: Done, removed the migration file

File: src/modules/testimonials/testimonials.controller.ts [hunk 5]
```diff
@@ -38,4 +38,20 @@ export class TestimonialsController {
...
     };
   }
+
+  @ApiOperation({ summary: 'Delete Testimonial' })
+  @ApiResponse({ status: 200, description: 'Testimonial deleted successfully' })
+  @ApiResponse({ status: 404, description: 'Testimonial not found' })
+  @ApiResponse({ status: 403, description: 'Access denied' })
+  @ApiResponse({ status: 500, description: 'Internal Server Error' })
+  @Delete(':id')
+  async deleteTestimonial(@Param('id', ParseUUIDPipe) id: string) {
```
[the remark this answers is not in the document]
Comment (Daniel-Ojo-Williams) [reply src/modules/testimonials/testimonials.controller.ts:48]: Only authenticated users can use this endpoint, the auth guard ensures that. The only thing remaining is to ensure that the authenticated user has the right permission or role.

File: src/modules/testimonials/testimonials.service.ts (hunk 3 above)
Comment (Daniel-Ojo-Williams) [reply src/modules/testimonials/testimonials.service.ts:63]: I don't think we'll want to delete a testimonial if the user account was deleted.

Comment (CFCIfe) [reply src/modules/testimonials/testimonials.service.ts:63]: @Heba-WebDev and @TheCodeGhinux will be able to help us. But technically we should not be displaying a user's review when they deleted.

**Comment by CFCIfe** [comment]
Everything looks good here to me.

**Pull request closed without merging**